sdk-lib/src/main/java/cash/z/ecc/android/sdk/block/CompactBlockProcessor.kt

```diff
@@ -426,30 +477,35 @@ class CompactBlockProcessor(
     }
 
     var failedUtxoFetches = 0
-    internal suspend fun refreshUtxos(tAddress: String, startHeight: Int): Int? = withContext(IO) {
-        var count: Int? = null
-        // todo: cleanup the way that we prevent this from running excessively
-        //       For now, try for about 3 blocks per app launch. If the service fails it is
-        //       probably disabled on ligthtwalletd, so then stop trying until the next app launch.
-        if (failedUtxoFetches < 9) { // there are 3 attempts per block
-            try {
-                retryUpTo(3) {
-                    val result = downloader.lightWalletService.fetchUtxos(tAddress, startHeight)
-                    count = processUtxoResult(result, tAddress, startHeight)
+    internal suspend fun refreshUtxos(tAddress: String, startHeight: BlockHeight): Int? =
+        withContext(IO) {
+            var count: Int? = null
+            // todo: cleanup the way that we prevent this from running excessively
+            //       For now, try for about 3 blocks per app launch. If the service fails it is
+            //       probably disabled on ligthtwalletd, so then stop trying until the next app launch.
+            if (failedUtxoFetches < 9) { // there are 3 attempts per block
+                try {
+                    retryUpTo(3) {
+                        val result = downloader.lightWalletService.fetchUtxos(tAddress, startHeight)
+                        count = processUtxoResult(result, tAddress, startHeight)
+                    }
+                } catch (e: Throwable) {
+                    failedUtxoFetches++
+                    twig("Warning: Fetching UTXOs is repeatedly failing! We will only try about ${(9 - failedUtxoFetches + 2) / 3} more times then give up for this session.")
                 }
-            } catch (e: Throwable) {
-                failedUtxoFetches++
-                twig("Warning: Fetching UTXOs is repeatedly failing! We will only try about ${(9 - failedUtxoFetches + 2) / 3} more times then give up for this session.")
+            } else {
+                twig("Warning: gave up on fetching UTXOs for this session. It seems to unavailable on lightwalletd.")
             }
-        } else {
-            twig("Warning: gave up on fetching UTXOs for this session. It seems to unavailable on lightwalletd.")
+            count
         }
-        count
-    }
 
-    internal suspend fun processUtxoResult(result: List<Service.GetAddressUtxosReply>, tAddress: String, startHeight: Int): Int = withContext(IO) {
+    internal suspend fun processUtxoResult(
+        result: List<Service.GetAddressUtxosReply>,
+        tAddress: String,
+        startHeight: BlockHeight
+    ): Int = withContext(IO) {
         var skipped = 0
-        val aboveHeight = startHeight - 1
+        val aboveHeight = startHeight + -1
```

Comment:
is this kind of syntax expected on kotlin or would it be better to have a `decrement` helper for the type?

Reply:
This is a weird special case that invented today.  It is a kludge to support support the existing logic in CompactBlockProcessor, although it could probably have a further refactor to make this cleaner.

BlockHeight has an `add` operator, but no subtract operator.  That's to ensure that subtracting two BlockHeights don't violate the representation invariant of the height being above the sapling activation height for the network.  BlockHeights don't store the network, so a subtraction can't know whether the result is valid.  (We only validate when using the `BlockHeight.Companion.new` factory method).  Because I intentionally left out the subtract operator, I'm doing plus -1 to effectively do a subtract.  The add operator allows this, but doesn't allow plus -2 since the only places we ever subtract was at most -1.

I don't like this solution, but it doesn't leak into the public API so it is mostly not too bad.

Two alternatives are
 - Do a second pass to update CompactBlockProcessor to change the bounds so that -1 isn't necessary
 - Keep the network inside the BlockHeight, so that subtraction can be validated.  (I don't like this option).

Thoughts?

Reply:
> Do a second pass to update CompactBlockProcessor to change the bounds so that -1 isn't necessary

If the subtractions are bound to the CompactBlockProcessor, then probably we can be fine with BlockHeight not needing any math and refactor the processor, which is the class introducing this requirement. Do you have a ballpark estimate of what the refactor would be?

Reply:
I'll have to scope this out and follow up.

There is at least one place where I think this impacts the Rust code, due to inclusive versus exclusive boundary conditions.  So in that particular case, changing it would be more difficult.

Reply:
So I did a few things to make this better.  It isn't 100% but its most of the way there.